lib/Controller/SocialApiController.php

```diff
@@ -0,0 +1,188 @@
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Contacts\Controller;
+
+use OCP\AppFramework\ApiController;
+use OCP\AppFramework\Http\TemplateResponse;
+// use OCP\IInitialStateService;
+use OCP\IConfig;
+use OCP\L10N\IFactory;
+use OCP\IRequest;
+
+class SocialApiController extends ApiController {
+
+	protected $appName;
+
+	// /** @var IInitialStateService */
+	// private $initialStateService;
+
+	/** @var IFactory */
+	private $languageFactory;
+	/** @var IConfig */
+	private  $config;
+
+	public function __construct(string $AppName,
+								IRequest $request,
+								IConfig $config,
+								// IInitialStateService $initialStateService,
+								IFactory $languageFactory) {
+		parent::__construct($AppName, $request);
+
+		$this->appName = $AppName;
+		// $this->initialStateService = $initialStateService;
+		$this->languageFactory = $languageFactory;
+		$this->config = $config;
+	}
+
+
+	/**
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
+	 *
+	 * generate download url for a social entry (based on type of data requested)
+	 *
+	 * @param {array} socialentry entry of contact
+	 * @param string type which information to link to (avatar, ...)
+	 * @return string
+	 */
+	protected function getSocialConnector($socialentry, $type) : ?string {
+		if (!is_array($socialentry)) {
+			throw new Exception("socialentry format missmatch"); // FIXME: the Exceptions seem not to work as expected...
+		}
+
+		$candidate = $socialentry[3];
+		$network   = $socialentry[1]['type'];
+		$connector = null;
+
+		if (is_array($network)) { $network = $network[0]; }
+
+		// get profile-id
+		switch ($network) {
+			case "facebook":
+				$candidate = basename($candidate);
+				if (!ctype_digit($candidate)) {
+					// TODO: determine facebook profile id from username
+					throw new Exception("facebook profile-id expected to be a number, not %s", $candidate);
+				}
+				break;
+			default:
+				throw new Exception("%s not implemented", $network);
+		}
+
+		// build connector
+		switch ($network) {
+			case "facebook":
+				switch ($type) {
+					case "avatar":
+						$connector = "https://graph.facebook.com/" . ($candidate) . "/picture?width=720";
+						break;
+					default:
+						throw new Exception("%s for %s not implemented", $type, $network);
+				}
+				break;
+			default:
+				throw new Exception("Unexpected error building the connector for %s", $network);
+		}
+
+		return ($connector);
+
+	}
+
+	/**
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
+	 *
+	 * Retrieves social profile data for a contact
+	 *
+	 * @param string addressbookId identifier of the addressbook
+	 * @param string contact identifier of the contact
+	 * @param string type which information to get: avatar, ...
+	 */
+	public function fetch($addressbookId, $contactId, $type) {
+
+		$url = null;
+		$response = 404;
+
+		try {
+			// get social parameters from contact
+
+			/* TODO - port from javascript...
+			const jCal = this.contact.jCal
```

Comment:
So, to get the contact in here, we have an interface for it: https://github.com/nextcloud/server/blob/a8a06a82d2b93c17b3014f0d468f23239100ca10/lib/public/Contacts/IManager.php#L176

Reply:
Then you should be able to parse the contacts info and retrieve the appropriate X-SOCIALPROFILE data :)

Reply:
Perfect, that helped.
But discovered either a bug or a feature:

This is how I access the contact data now:

```
$contact = $this->manager->search($contactId, array('UID'))[0];
$socialprofile = $contact['X-SOCIALPROFILE'];
```

Which works fine. But when there are more than one id for a network, this is stored fine in the VCARD but lost in my search. What am I missing? This is the VCard of the contact:

```
BEGIN:VCARD
VERSION:3.0
PRODID:-//Sabre//Sabre VObject 4.1.6//EN
UID:3225c0d5-1bd2-43e5-a08c-4e65eaa406b0
REV;VALUE=DATE-AND-OR-TIME:20200416T222255Z
FN:Mr. Example
ADR;TYPE=HOME:;;;;;;
EMAIL;TYPE=HOME:
X-SOCIALPROFILE;TYPE=facebook:https://facebook.com/5
X-SOCIALPROFILE;TYPE=FACEBOOK:4
X-SOCIALPROFILE;TYPE=facebook:666
PHOTO;ENCODING=b:a29udGFrdGU8YnI+ZG9uZTxicj4=
END:VCARD
```

Reply:
Hey!
Nice catch! This is indeed a bug: https://github.com/nextcloud/server/blob/caff1023ea72bb2ea94130e18a2a6e2ccf819e5f/apps/dav/lib/CardDAV/AddressBookImpl.php#L274

Btw be careful because you can have the same UID across multiple addressbooks, so if you take the first of the array, you might not get the one you want :)

I'll provide a fix for server :wink:

Reply:
Fix in https://github.com/nextcloud/server/pull/20529
Feel free to review :)

Reply:
```php
		$addressBooks = $this->manager->getUserAddressBooks();
		$addressBook = null;
		foreach($addressBooks as $ab) {
			// searching the addressbook id 'contacts'
			if ($ab->getUri() === 'contacts') {
				$addressBook = $ab;
			}
		}
		// searching the contact uid '0c3e6594-7a4a-4940-9248-5c8ee3e23b1f'
		$results = $addressBook->search('0c3e6594-7a4a-4940-9248-5c8ee3e23b1f', ['UID'], ['types' => true]);

```

Reply:
Wow, that was a quick fix! Awesome

Reply:
Unfortunately I still could not figure out how to update the contact. I'm starting to think it's a [bug](https://github.com/nextcloud/server/issues/20544)...

Reply:
So, looking at other examples, you can also do it diffrently, which is probably better actually
https://github.com/nextcloud/maps/blob/ea23192efae5eeab51d66ad91cce9017ef2c4d74/lib/Controller/ContactsController.php#L189-L193

Use the CardDavBackend service instead of the AddressBook search and createOrUpdate

Reply:
Thank you for your help! By looking at the source code of the update-function, I finally discovered that I can avoid the issue here by setting UID and PHOTO only. Phew!